consensus/byzantine_test.go

```diff
@@ -3,32 +3,152 @@ package consensus
+
+		blockDB := dbm.NewMemDB()
+		blockStore := store.NewBlockStore(blockDB)
+
+		// one for mempool, one for consensus
+		mtx := new(tmsync.Mutex)
+		proxyAppConnMem := abcicli.NewLocalClient(mtx, app)
+		proxyAppConnCon := abcicli.NewLocalClient(mtx, app)
+
+		// Make Mempool
+		mempool := mempl.NewCListMempool(thisConfig.Mempool, proxyAppConnMem, 0)
+		mempool.SetLogger(log.TestingLogger().With("module", "mempool"))
+		if thisConfig.Consensus.WaitForTxs() {
+			mempool.EnableTxsAvailable()
+		}
+
+		// Make a full instance of the evidence pool
+		evidenceDB := dbm.NewMemDB()
+		evpool, err := evidence.NewPool(stateDB, evidenceDB, blockStore)
+		require.NoError(t, err)
+		evpool.SetLogger(logger.With("module", "evidence"))
+
+		// Make State
+		blockExec := sm.NewBlockExecutor(stateDB, log.TestingLogger(), proxyAppConnCon, mempool, evpool)
+		cs := NewState(thisConfig.Consensus, state, blockExec, blockStore, mempool, evpool)
+		cs.SetLogger(cs.Logger)
+		// set private validator
+		pv := privVals[i]
+		cs.SetPrivValidator(pv)
+
+		eventBus := types.NewEventBus()
+		eventBus.SetLogger(log.TestingLogger().With("module", "events"))
+		eventBus.Start()
+		cs.SetEventBus(eventBus)
+
+		cs.SetTimeoutTicker(tickerFunc())
+		cs.SetLogger(logger)
+
+		css[i] = cs
+	}
+
+	reactors, blocksSubs, eventBuses := startConsensusNet(t, css, nValidators)
+	defer stopConsensusNet(log.TestingLogger(), reactors, eventBuses)
+
+	// create byzantine validator
+	bcs := css[0]
+
+	bcs.doPrevote = func(height int64, round int32) {
+		// allow first height to happen normally so that byzantine validator is no longer proposer
+		if height == 2 {
+			bcs.Logger.Info("Sending two votes")
+			// bcs.mtx.Lock()
+			prevote1, err := bcs.signVote(tmproto.PrevoteType, bcs.ProposalBlock.Hash(), bcs.ProposalBlockParts.Header())
+			require.NoError(t, err)
+			prevote2, err := bcs.signVote(tmproto.PrevoteType, nil, types.PartSetHeader{})
+			require.NoError(t, err)
+			// bcs.mtx.Unlock()
+			peerList := reactors[1].Switch.Peers().List()
+			bcs.Logger.Info("Getting peer list", "peers", peerList)
+			// send two votes to all peers
+			for i, peer := range peerList {
+				if i < len(peerList)/2 {
+					bcs.Logger.Info("Signed and pushed vote", "vote", prevote1, "peer", peer)
+					peer.Send(VoteChannel, MustEncode(&VoteMessage{prevote1}))
+				} else {
+					bcs.Logger.Info("Signed and pushed vote", "vote", prevote2, "peer", peer)
+					peer.Send(VoteChannel, MustEncode(&VoteMessage{prevote2}))
+				}
+			}
+		} else {
+			bcs.Logger.Info("Behaving normally")
+			bcs.defaultDoPrevote(height, round)
+		}
+	}
+
+	// Check that evidence is submitted and committed at the third height
+	for i := 0; i < 2; i++ {
+		timeoutWaitGroup(t, nValidators, func(j int) {
+			<-blocksSubs[j].Out()
+		}, css)
+	}
+	timeoutWaitGroup(t, nValidators, func(j int) {
+		msg := <-blocksSubs[j].Out()
+		block := msg.Data().(types.EventDataNewBlock).Block
+		assert.True(t, len(block.Evidence.Evidence) == 1)
```

Comment:
can we check the evidence type?

Reply:
I also now check that the addresses match